### PR TITLE
JP-2664: Fix NRS-MOS source ids to be 5 digits and positive

### PR DESCRIPTION
**Pull request by tapastro**
Closes #
Resolves [JP-2664](https://jira.stsci.edu/browse/JP-2664)

**Description**

This PR adds a check on the source_id created in calwebb_spec3, and modifies values that fall outside 0-99999. The wording in the JP ticket indicated this is a temporary fix, so it has been worded/coded as such; if I misinterpreted the level of 'hotfix' involved, I can edit the addition.

Checklist
- [ ] Tests
- [ ] Documentation
- [x] Change log
- [x] Milestone
- [x] Label(s)

### CODE REVIEW

File: jwst/pipeline/calwebb_spec3.py [hunk 1]
```diff
@@ -162,6 +163,32 @@ def process(self, input):
...
+                available_src_ids.remove(src_id)
+
+        hotfixed_sources = []
+        # now find and reset bad source_id values
+        for src in sources:
+            src_id, model = src
+            src_id = int(src_id)
+            if src_id < 0 or src_id > 99999:
+                src_id = available_src_ids.pop()
+            hotfixed_sources.append((str(src_id), model))
```
[the remark this answers is not in the document]
Comment (nden) [reply jwst/pipeline/calwebb_spec3.py:188]: Just trying to understand - originally the `source_id` value comes from the metadata conf file when it's read in assign_wcs. Source_id for background slits is assigned a number greater than the max_source_id, so it's likely those that need adjustment are background slitlets. 
Then `source_id` is assigned to an extracted SlitModel in extract2d based on what was assigned to a slitlet in assign_wcs. So if this is correct shouldn't the fix be in assign_wcs?

[the remark this answers is not in the document]
Comment (nden) [reply jwst/pipeline/calwebb_spec3.py:188]: Never mind, should have read the ticket first.

**Comment by tapastro** [comment]
@hbushouse Addressed the source_id listed in the datamodels - should there be a log message mapping old to new source_ids?

**Comment by hbushouse** [comment]
> @hbushouse Addressed the source_id listed in the datamodels - should there be a log message mapping old to new source_ids?

Yes, I think a log message would be nice to have.

**Comment by hbushouse** [comment]
> > Some test data files are in the directory listed in JP-2664. None of them have any negative source_ids, but almost all of them are > 99999. So that would make a good test to see the new "sxxxxx" fields in the output file names and SOURCEID values in the headers.
> 
> I had data from the same PID with negative values (which spawned this ticket), and I find similar results as found for the >5 digit numbers - source_ids are created starting from 00001 and work their way up. Loaded datamodels show matching values. Results are in the ticket directory (for sources until I lost VPN connection and broke the processing run).

Yeah, the earlier version of the data that you had was created using a bad MSA meta file that had the source_id column of the shutters table defined as int16 instead of int32. So all of the source_ids bigger than the int16 limit rolled over to negative numbers. The ones in the data directory now were reprocessed using a corrected MSA metafile from Mike, which made all the negative values go away. But legit negative values can still occur for slits that don't have a real source from the catalog.

**Comment by hbushouse** [comment]
> Is the reason to limit to 5 digits only to prevent complications down the line, specifically archive? Just curious.

Yes. I believe the archive ingest is setup to expect 5, and only 5, digits in the source id field of the file names and I don't think anyone has checked in detail yet to know whether Cal is OK with creating ones that have more than 5 digits. Haven't actually checked the part(s) of the code yet that fill in the {source_id} part of product names somewhere in level 3 processing. Do you happen to recall where that's done?

**Comment by hbushouse** [comment]
Ah hah. Found it: https://github.com/spacetelescope/jwst/blob/master/jwst/associations/lib/rules_level3_base.py#L621  calwebb_spec3 calls `format_product`, which definitely has the source_id field in the product file names hardwired to 5 digits.

**Review by hbushouse — APPROVED** [review]
Code looks good. Results look good (including the internal SOURCEID keyword values). I approve.

**Pull request merged**